### ISSUE TITLE
[2.0] Revise configuration approach

### ISSUE THREAD

**Issue by colinodell** [thephpleague/commonmark#447]
Our approach to configuration has a few shortcomings:

 - No support for providing default values, overriding them during MD->HTML conversion, and then rolling back before subsequent conversions
 - No support for validating types
 - Too many ways to set, get, and override configurations, some of which support `/`-delimited paths and others that don't
 - `ConfigurationAwareInterface` potentially allows people to pull configuration values out immediately instead of hanging onto the `ConfigurationInterface` object and lazily checking values later - the latter is recommended, but we don't document that anywhere.
 - It's basically a free-for-all

The current approach works well enough for 1.x, but it's going to prevent some features like #442 being able to override configuration values (without causing weird side effects)

I'd therefore like to revisit how we implement, handle, and document Configuration within this library for v2.

**Comment by markcarver** [comment]
There are plenty of "config" based packages already out there. Here's a list based on https://packagist.org/?query=config and the "features" we'd likely want/need. So far, `symfony/config` seems like maybe the best fit (very type oriented) if we really don't care about exporting/importing from files. If none of these are suitable, we'd still likely need to create our own implementation that handles nearly most of these features (for DX and extensibility purposes).

| Feature                   | [symfony/config] | [symfony/options-resolver] | [hassankhan/config] | [dflydev/dot-access-data] | [laminas/laminas-config] |
|:--------------------------|:----------------:|:--------------------------:|:-------------------:|:-------------------------:|:------------------------:|
| Dependencies              | 4                | 2                          |                     |                           | 3                        |
| Nested config/hierarchy   | ✓                | ✓                          | ✓                   | ✓                         | ✓                        |
| Define default values     | ✓                | ✓                          | via `get()` fallback|                           |                          |
| Define deprecations       | ✓                | ✓                          |                     |                           |                          |
| Define required values    | ✓                | ✓                          |                     |                           |                          |
| Description (docs)        | ✓                |                            |                     |                           |                          |
| Get value: array access   | ✓                | ✓                          | ✓                   |                           | ✓                        |
| Get value: object/method  | ✓                |                            | ✓                   |                           | ✓                        |
| Get value: dot-notation   | unclear/possible |                            | ✓                   |                           |                          |
| Immutabilty               |                  |                            |                     |                           | ✓                        |
| Merging                   | ✓                |                            | ✓                   | ✓ via `import()`          | ✓                        |
| Read from file (INI)      |                  |                            | ✓                   |                           | ✓                        |
| Read from file (JSON)     |                  |                            | ✓                   |                           | ✓                        |
| Read from file (PHP/Array)|                  |                            | ✓                   |                           |                          |
| Read from file (XML)      |                  |                            | ✓                   |                           | ✓                        |
| Read from file (YAML)     |                  |                            | ✓                   |                           | ✓                        |
| Set value: array access   | ✓                | ✓                          | ✓                   |                           | ✓                        |
| Set value: object/method  | ✓                |                            | ✓                   | ✓                         | ✓                        |
| Set value: dot-notation   | unclear/possible |                            | ✓                   | ✓                         |                          |
| Types                     | ✓                | ✓                          |                     |                           |                          |
| Type validation           | ✓                | ✓                          |                     |                           | ✓ via convoluted processor |
| Value normalization       | ✓                | ✓                          |                     |                           | ✓ via convoluted processor |
| Value validation          | ✓                | ✓                          |                     |                           | ✓ via convoluted processor |
| Write to file (INI)       |                  |                            | ✓                   |                           | ✓                        |
| Write to file (JSON)      |                  |                            | ✓                   |                           | ✓                        |
| Write to file (PHP/Array) |                  |                            | ✓                   |                           | ✓                        |
| Write to file (XML)       |                  |                            | ✓                   |                           | ✓                        |
| Write to file (YAML)      |                  |                            | ✓                   |                           | ✓                        |

[symfony/config]: https://packagist.org/packages/symfony/config
[symfony/options-resolver]: https://packagist.org/packages/symfony/options-resolver
[dflydev/dot-access-data]: https://packagist.org/packages/dflydev/dot-access-data
[hassankhan/config]: https://packagist.org/packages/hassankhan/config
[laminas/laminas-config]: https://packagist.org/packages/laminas/laminas-config

**Comment by markcarver** [comment]
Is this still a 2.0 issue? Should we create a new 3.0 milestone for this?

**Comment by colinodell** [comment]
I would like to include this in 2.0.

After tinkering with the different libraries above and doing some additional research, I think I've refined the type of approach that would be best.  In a nutshell, we need to separate the config functionality into three distinct parts:

## 1. Defining the configuration schema

When extensions are being registered, they can define any configuration names/keys, default values, validation, etc. that they may want.  This would be done within their `register()` method by calling a method like `$environment->addConfigSchema(string $key, Schema $schema)`.

Only the schema (which includes defaults) can be set at this time, and they can never be modified outside of this.

## 2. Setting the user's desired configuration values

We'll combine `ConfigurableEnvironmentInterface::mergeConfig()` and `::setConfig()` into a single method which the user can call when setting up their environment.  They would pass in all their desired configuration values here.  (These values won't be validated or "go into effect" just yet, since extensions are registered lazily - we just store them temporarily)

This can be done anytime before parsing/rendering beings.

## 3. "Finalizing" the configuration for a single conversion execution

Immediately before the parsing step, we'll apply the user's configuration values against the schema.  If there's a validation error we'll throw an exception.  Otherwise, we'll now have a valid configuration we can use for parsing and rendering that single document.

However, we need to provide one final opportunity for the configuration values to be set at the very start of the parsing/rendering.  This is primarily needed to allow for [front matter to adjust the configuration](https://github.com/thephpleague/commonmark/issues/442) for the current document.  It would look at the current configuration, see if that set-config-based-on-front-matter config value is enabled, and adjust the configuration if so.

The biggest challenge here is that `MarkdownParser` and `HtmlRenderer` only have a single shared resource between them - the `Environment`.  But we can't feasibly store the one-off configuration in there because those values would then apply to subsequent documents, which is the problem we have today.

I'd therefore propose adding in some kind of new `ExecutionContext` object (better name TBD) which only exists for the lifetime of the `convertToHtml()` call and shares the following items between the parser engine and renderer engine:

 - The `Environment`
 - The "finalized" configuration values unique to that one document

Instead of individual parser/renderer/listener/etc classes implementing `ConfigurationAware` and/or `EnvironmentAwareInterface`, there would be a single `ContextAwareInterface` they would receive which provides access to both.  Implementors should expect this context to be provided at the very last minute, and possibly multiple different instances for each new document being parsed.

This `ExecutionContext` could also provide arbitrary data storage.  Right now, things like the Front Matter extension are (ab)using the `Document`'s `$data` property to store data and make it available to other things, which is more of a kludgy workaround than an ideal scenario.

## Backward Compatibility

I'd imagine the following breaking changes would be made to existing classes and methods:

| Class | Method | Change |
| --- | --- | --- |
| `ConfigurableEnvironmentInterface` | `addConfigSchema()` | New method |
| `ConfigurableEnvironmentInterface` | `setConfig()` | Remove (but `mergeConfig()` would likely remain) |
| `EnvironmentInterface` | `getConfig()` | Remove |
| `MarkdownParser` | `__construct()` | Change first parameter from `EnvironmentInterface` to the new `ExecutionContext` or whatever we call it |
| `HtmlRenderer` | `__construct()` | Change first parameter from `EnvironmentInterface` to the new `ExecutionContext` or whatever we call it |
| `EnvironmentAwareInterface` | | Remove this interface |
| `ConfigurationAwareInterface` | | Remove this interface |

## Implementation Details

Of all the different libraries, I think `symfony/options-resolver` and `nette/schema` are the two closest matches to what we'd want.  They both support defining the various options with types, defaults, validation, etc. and then applying an array against that to get the final configuration.  Neither has support for getting/setting values by dot-notation but we can easily overcome that by wrapping the final array in our own class which exposes that functionality.

To me, the biggest difference between the two libraries that we'd care about comes down to:

| Feature | `symfony/options-resolver` | `nette/schema` |
|:------------------------------|:--------------:|:--------------------------:|
| Dependencies | 2 | 1 |
| Validation | Fixed values only; or roll-your-own | ✓ |
| Define deprecations | ✓ | |
| Expressive interface for defining options | | ✓ |
| Human-friendly descriptions | | |

Weighing those options, I'm leaning more towards using `nette/schema` (or perhaps even forking it to add/remove certain features to better suit our wishlist)

**Comment by markcarver** [comment]
Sounds good to me.

**Comment by colinodell** [comment]
Upon further consideration, I don't think my `ExecutionContext` idea will work exactly as described above, but the high-level ideas in those three bullet points should still work.  I'm playing with a few different approaches now and should hopefully have some working code soon!

**Comment by markcarver** [comment]
FWIW, I tried using `nette/schema` with the emoji stuff and had a hard time finding it "useful". Ended up just going with `symfony/options-resolver` and `dflydev/dot-access-data`. Then just created a simple "wrapper" `Configuration` class:

https://github.com/unicorn-fail/emoji/blob/latest/src/Configuration.php

**Comment by colinodell** [comment]
I generally prefer the Symfony library but after playing with `nette/schema` I really like their separation of concerns and how the schema is an immutable object.  This means each extension can describe the schema it wants to use and we don't have to worry about the extension changing the schema later on or grabbing options out prematurely.

I'm also debating forking `dflydev/dot-access-data`.  There are some features I want to add and change which can't easily be done without breaking changes.

**Comment by colinodell** [comment]
I've been working with the maintainer of `dflydev/dot-access-data` to add the features we need to their next major release :)

**Comment by markcarver** [comment]
Are you still going to fork that project to add on top of it the other things that are needed (i.e. defaults, required, deprecations, type validation, etc.)?

I only ask because the dot-notation, which is a huge DX improvement, should really just be the sugar on top of the underlying config structure that's actually needed.

**Comment by markcarver** [comment]
Or rather, were you planning on using both `symfony/options-resolver` and `dflydev/dot-access-data` 
 to do something similar as I mentioned in https://github.com/thephpleague/commonmark/issues/447#issuecomment-692138083

**Comment by colinodell** [comment]
I plan on using `nette/schema` for defining the configuration options (names, types, defaults, validation, etc) plus `dflydev/dot-access-data` (upcoming 3.0 release) for accessing the final ("resolved") option values.

Extensions will define their configuration schema like this:

```php
final class MentionExtension implements ConfigurableExtensionInterface
{
    public function configureSchema(ConfigurationBuilderInterface $builder): void
    {
        $isAValidPartialRegex = static function (string $regex): bool {
            $regex = '/' . $regex . '/i';

            return @\preg_match($regex, '') !== false;
        };

        $builder->addSchema('mentions', Expect::arrayOf(
            Expect::structure([
                'prefix' => Expect::string()->required(),
                'pattern' => Expect::string()->assert($isAValidPartialRegex, 'Pattern must not include starting/ending delimiters (like "/")')->required(),
                'generator' => Expect::anyOf(
                    Expect::type(MentionGeneratorInterface::class),
                    Expect::string(),
                    Expect::type('callable')
                )->required(),
            ])->castTo('array')
        ));
    }

    public function register(ConfigurableEnvironmentInterface $environment): void
    {
        // ...
    }
}
```

I'm waiting for feedback on https://github.com/nette/schema/issues/27 to determine whether/how support for deprecated options might be allowed.

**Comment by markcarver** [comment]
Sounds good to me. I may replace the emoji configuration with the same once it's all finalized here just its similar.

**Comment by colinodell** [comment]
Implemented via #598
